File: server/src/schema/tribulation.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TribulationKindV1 {
    DuXu,
    ZoneCollapse,
    Targeted,
    AscensionQuotaOpen,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case", tag = "kind")]
pub enum TribulationPhaseV1 {
    Omen,
    Lock,
    Wave { wave: u32 },
    HeartDemon,
    Settle,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DuXuOutcomeV1 {
    Ascended,
    HalfStep,
    Failed,
    Killed,
    Fled,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DuXuResultV1 {
    pub char_id: String,
    pub outcome: DuXuOutcomeV1,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub killer: Option<String>,
    pub waves_survived: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct TribulationEventV1 {
    pub v: u8,
    pub kind: TribulationKindV1,
    pub phase: TribulationPhaseV1,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub char_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub actor_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub zone: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub epicenter: Option<[f64; 3]>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub wave_current: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub wave_total: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<DuXuResultV1>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub occupied_slots: Option<u32>,
}

impl TribulationEventV1 {
    pub fn du_xu(
        phase: TribulationPhaseV1,
        char_id: Option<String>,
        actor_name: Option<String>,
        epicenter: Option<[f64; 3]>,
        wave_current: Option<u32>,
        wave_total: Option<u32>,
        result: Option<DuXuResultV1>,
    ) -> Self {
        Self {
            v: 1,
            kind: TribulationKindV1::DuXu,
            phase,
            char_id,
            actor_name,
            zone: None,
            epicenter,
            wave_current,
            wave_total,
            result,
            occupied_slots: None,
        }
    }

    pub fn zone_collapse(
        phase: TribulationPhaseV1,
        zone: Option<String>,
        epicenter: Option<[f64; 3]>,
    ) -> Self {
        Self {
            v: 1,
            kind: TribulationKindV1::ZoneCollapse,
            phase,
            char_id: None,
            actor_name: None,
            zone,
            epicenter,
            wave_current: None,
            wave_total: None,
            result: None,
            occupied_slots: None,
        }
    }

    pub fn targeted(
        phase: TribulationPhaseV1,
        zone: Option<String>,
        epicenter: Option<[f64; 3]>,
    ) -> Self {
        Self {
            v: 1,
            kind: TribulationKindV1::Targeted,
            phase,
            char_id: None,
            actor_name: None,
            zone,
            epicenter,
            wave_current: None,
            wave_total: None,
            result: None,
            occupied_slots: None,
        }
    }

    pub fn ascension_quota_open(occupied_slots: Option<u32>) -> Self {
        Self {
            v: 1,
            kind: TribulationKindV1::AscensionQuotaOpen,
            phase: TribulationPhaseV1::Settle,
            char_id: None,
            actor_name: None,
            zone: None,
            epicenter: None,
            wave_current: None,
            wave_total: None,
            result: None,
            occupied_slots,
        }
    }
}

File: agent/packages/schema/src/tribulation.ts
/**
 * plan-tribulation-v1 §6 天劫 IPC schema。与 `server/src/schema/tribulation.rs` 对齐。
 */
import { Type, type Static } from "@sinclair/typebox";
import { validate, type ValidationResult } from "./validate.js";

export const TribulationKindV1 = Type.Union([
  Type.Literal("du_xu"),
  Type.Literal("zone_collapse"),
  Type.Literal("targeted"),
  Type.Literal("jue_bi"),
  Type.Literal("ascension_quota_open"),
]);
export type TribulationKindV1 = Static<typeof TribulationKindV1>;

export const TribulationPhaseV1 = Type.Union([
  Type.Object({ kind: Type.Literal("omen") }, { additionalProperties: false }),
  Type.Object({ kind: Type.Literal("lock") }, { additionalProperties: false }),
  Type.Object(
    {
      kind: Type.Literal("wave"),
      wave: Type.Integer({ minimum: 0 }),
    },
    { additionalProperties: false },
  ),
  Type.Object({ kind: Type.Literal("heart_demon") }, { additionalProperties: false }),
  Type.Object({ kind: Type.Literal("settle") }, { additionalProperties: false }),
]);
export type TribulationPhaseV1 = Static<typeof TribulationPhaseV1>;

export const DuXuOutcomeV1 = Type.Union([
  Type.Literal("ascended"),
  Type.Literal("half_step"),
  Type.Literal("failed"),
  Type.Literal("killed"),
  Type.Literal("fled"),
]);
export type DuXuOutcomeV1 = Static<typeof DuXuOutcomeV1>;

export const DuXuResultReasonV1 = Type.Union([
  Type.Literal("void_quota_exceeded"),
]);
export type DuXuResultReasonV1 = Static<typeof DuXuResultReasonV1>;

export const DuXuResultV1 = Type.Object(
  {
    char_id: Type.String({ minLength: 1 }),
    outcome: DuXuOutcomeV1,
    killer: Type.Optional(Type.String({ minLength: 1 })),
    waves_survived: Type.Integer({ minimum: 0 }),
    reason: Type.Optional(DuXuResultReasonV1),
  },
  { additionalProperties: false },
);
export type DuXuResultV1 = Static<typeof DuXuResultV1>;

export const TribulationEventV1 = Type.Object(
  {
    v: Type.Literal(1),
    kind: TribulationKindV1,
    phase: TribulationPhaseV1,
    char_id: Type.Optional(Type.String({ minLength: 1 })),
    actor_name: Type.Optional(Type.String({ minLength: 1 })),
    zone: Type.Optional(Type.String({ minLength: 1 })),
    source: Type.Optional(Type.String({ minLength: 1 })),
    epicenter: Type.Optional(Type.Tuple([Type.Number(), Type.Number(), Type.Number()])),
    wave_current: Type.Optional(Type.Integer({ minimum: 0 })),
    wave_total: Type.Optional(Type.Integer({ minimum: 0 })),
    result: Type.Optional(DuXuResultV1),
    occupied_slots: Type.Optional(Type.Integer({ minimum: 0 })),
  },
  { additionalProperties: false },
);
export type TribulationEventV1 = Static<typeof TribulationEventV1>;

export function validateTribulationEventV1Contract(data: unknown): ValidationResult {
  return validate(TribulationEventV1, data);
}

// ─── plan-halfstep-rechallenge-integration-v1 P1 ────────────────────────────

/**
 * 半步化虚重渡触发 payload（server → agent，`bong:tribulation/halfstep_rechallenge`）。
 *
 * - `char_id`：触发方玩家/NPC 的 char_id（agent 用于 player-scope narration target）
 * - `zone_name`：触发方当前所在 zone 名称（agent 用于 zone-scope narration target；
 *               dormant 触发方填 "dormant"）
 * - `zone_halfstep_count`：当前同 zone 内 HalfStep 修士总数（含触发方；
 *                          agent 据此决定是否 emit zone echo narration，门槛 >= 2）
 * - `at_tick`：server 游戏 tick（时序审计）
 */
export const HalfStepRechallengeTriggerPayloadV1 = Type.Object(
  {
    char_id: Type.String({ minLength: 1 }),
    zone_name: Type.String({ minLength: 1 }),
    zone_halfstep_count: Type.Integer({ minimum: 0 }),
    at_tick: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type HalfStepRechallengeTriggerPayloadV1 = Static<
  typeof HalfStepRechallengeTriggerPayloadV1
>;

export function validateHalfStepRechallengeTriggerPayloadV1Contract(
  data: unknown,
): ValidationResult {
  return validate(HalfStepRechallengeTriggerPayloadV1, data);
}
